cctools-2/ar/ar.5
.\"	$OpenBSD: ar.5,v 1.2 1996/06/26 05:31:17 deraadt Exp $
.\"	$NetBSD: ar.5,v 1.2 1995/03/25 06:39:38 glass Exp $
.\"
.\" Copyright (c) 1990, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)ar.5.5	8.1 (Berkeley) 6/9/93
.\"
.Dd June 9, 1993
.Dt AR 5
.Os
.Sh NAME
.Nm ar
.Nd archive (library) file format
.Sh SYNOPSIS
.Fd #include <ar.h>
.Sh DESCRIPTION
The archive command
.Nm ar
combines several files into one.
Archives are mainly used as libraries of object files intended to be
loaded using the link-editor
.Xr ld 1 .
.Pp
A file created with
.Nm ar
begins with the ``magic'' string "!<arch>\en".
The rest of the archive is made up of objects, each of which is composed
of a header for a file, a possible file name, and the file contents.
The header is portable between machine architectures, and, if the file
contents are printable, the archive is itself printable.
.Pp
The header is made up of six variable length
.Tn ASCII
fields, followed by a
two character trailer.
The fields are the object name (16 characters), the file last modification
time (12 characters), the user and group id's (each 6 characters), the file
mode (8 characters) and the file size (10 characters).
All numeric fields are in decimal, except for the file mode which is in
octal.
.Pp
The modification time is the file
.Fa st_mtime
field, i.e.,
.Dv CUT
seconds since
the epoch.
The user and group id's are the file
.Fa st_uid
and
.Fa st_gid
fields.
The file mode is the file
.Fa st_mode
field.
The file size is the file
.Fa st_size
field.
The two-byte trailer is the string "\`\en".
.Pp
Only the name field has any provision for overflow.
This is not yet used by the Rhapsody compiler tools however.
If any file name is more than 16 characters in length or contains an
embedded space, the string "#1/" followed by the
.Tn ASCII
length of the
name is written in the name field.
The file size (stored in the archive header) is incremented by the length
of the name.
The name is then written immediately following the archive header.
.Pp
Any unused characters in any of these fields are written as space
characters.
If any fields are their particular maximum number of characters in
length, there will be no separation between the fields.
.Pp
Objects in the archive are always an even number of bytes long; files
which are an odd number of bytes long are padded with a newline (``\en'')
character, although the size in the header does not reflect this.
.Sh SEE ALSO
.Xr ar 1 ,
.Xr stat 2
.Sh HISTORY
There have been at least four
.Nm ar
formats.
The first was denoted by the leading ``magic'' number 0177555 (stored as
type int).
These archives were almost certainly created on a 16-bit machine, and
contain headers made up of five fields.
The fields are the object name (8 characters), the file last modification
time (type long), the user id (type char), the file mode (type char) and
the file size (type unsigned int).
Files were padded to an even number of bytes.
.Pp
The second was denoted by the leading ``magic'' number 0177545 (stored as
type int).
These archives may have been created on either 16 or 32-bit machines, and
contain headers made up of six fields.
The fields are the object name (14 characters), the file last modification
time (type long), the user and group id's (each type char), the file mode
(type int) and the file size (type long).
Files were padded to an even number of bytes.
For more information on converting from this format see
.Xr arcv 8 .
.Pp
The current archive format (without support for long character names and
names with embedded spaces) was introduced in
.Bx 4.0 .
The headers were the same as the current format, with the exception that
names longer than 16 characters were truncated, and names with embedded
spaces (and often trailing spaces) were not supported.
It has been extended for these reasons,
as described above.
This format first appeared in 4.4BSD.
.Sh COMPATIBILITY
No archive format is currently specified by any standard.
.At V
has historically distributed archives in a different format from
all of the above.


objc-1/ChangeLog
Changes from Ali to all the string class files. (12/23/92)
Changes from Derek to deal with name conflicts.

Cleaned up some warnings (7/23/91):

	* Makefile (COMMONFLAGS):  Changed -Wimplicit to -Wall.
	Let's make this compile	without warnings!

	* List.m, Storage.m (read:, write:):  Changed to return self.

	* <various files>:  Removed all unused local variables.

Added support for string objects (7/23/91):

	* NXString.m, NXCharacterSet.m, unichar.h, StringTest.m:  New files.

	* Makefile (LOCAL_CLASSES):  Added NXString.m and NXCharacterSet.m

	* Makefile (LOCAL_INCFILES):  Added unichar.h

	* objc-runtime.m (_objc_fixup_string_object_class):  New routine.
	If NXConstantString is present in class hash table, copy it to
	_NXConstantString_class and insert this address in class hashtable.

	* objc-runtime.m (_objcInit):  Call new routine after classes have
	been fixed up (in _objc_map_selectors).

	* objc-globaldata.m (_NXConstantString_class):  New global variable.
	Reduce pad by 36 bytes to compensate.

	* spec_sys:  Added slots for NXCompareCharacters(), NXEmptyString(),
	and NXFindCharacters().  Added files NXString.o and NXCharacterSet.o
	These changes will need to be integrated into libsys's copy.

Changed -[Object copy] (7/24/91):

	* Object.m (copy):  Chnaged to return [self copyFromZone: [self zone]]
	rather than using _copy.  The function object_copy() is no longer used
	by the runtime system, and could probably be removed.

	* List.[hm] (copy):  Deleted redundant method.

	* Storage.[hm] (copy):  Deleted redundant method.

	* HashTable.[hm] (copy):  Deleted redundant method.

	* NXString.[hm] (copy):  Deleted redundant method.

	* NXCharacterSet.[hm] (copy):  Deleted redundant method.

Changed method names in Storage (7/24/91):

	* Storage.[hm] (replaceElementAt:with:)  Changed from replace:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (insertElement:at:)  Changed from insert:at:
	and added 2.0 compatibility method in implementation only.

	* Storage.[hm] (removeElementAt:)  Changed from removeAt:
	and added 2.0 compatibility method in implementation only.

Cleaned up Makefile slightly (7/24/91):

	* Makefile (BY_PRODUCTS):  Added objcopt so that make clean removes
	all files not installed by make installsrc.

Fixed bug #14923 in HashTable (7/24/91):

	* HashTable.m (nextState:key:value:):  Set key and value to NULL
	when returning NO.

Fixed bug #14309 in Storage (7/24/91):

	* Storage.m (setNumSlots:):  Changed incorrect test
	(numSlots > maxElements) to (maxElements > old).

	* Storage.m (setNumSlots:):  Changed to use a single realloc.

Fixed bug #5687 in hashtable (7/24/91):

	* hashtable.h (NXHashInsertIfAbsent):  Improved comment.

Added appendList: method (dmoore 7/24/91)

	* List.m, List.h added method.
	
Added maptable functionality (dmoore 7/24/91)

	* maptable.c maptable.m

Removed typedef for id when compiling Objective-C (7/26/91):

	* objc.h (id):  Do not define typedef for id when compiling
	Objective-C (when __OBJC__ is predefined).
	
Made Objective C check for duplicate classes and use the one in the application
	(dmoore 7/27/91)
	
	* objc-runtime.m:  Added following functions:
		_objc_inform:	Just prints message for now.
		_resolveClassConflict: Makes sure that old class goes back in
			if it's from the app.
		_nameForHeader:	Returns the name of an image.
		_getExecHeader:	Picks out the app header.
		_headerForClass: Finds the mach_header containing a class.
		
		Modified _objc_get_classes to check for duplicates.
		
Added maptable prototypes to objc-globaltext.m (dmoore 7/27/91)

Modified spec_sys to include NXString and maptable stuff.

snaroff (7/31/91):

Integrate blaines kernel mods in a rational way.

Renamed "_objc_internal_error" to "_objc_inform"...removed "_objc_inform",
they were nearly identical (snaroff 7/31/91)

Create file "objc-errors.m"...it includes _objc_fatal, _objc_inform,
__objc_error, and _objc_error. MUST ADD _objc_inform and NXLogError to list 
of private externs!!!!!

Call _objc_fatal() instead of perror() in objc-sel.m...callings perror in these situations is bogus.

Have hashtable use NXLogError instead of fprintf(). 

Add field to objc_class for protocols...add protocol templates to objc-class.h

Installed new NXString files from ali (8/3/91):

	* NXString.[hm]:  Most methods are now statically typed.
	
	* NXStringTest.m  Changes to match.

Changes for clean compilation with new compiler (8/3/91):

	* Makefile (objcopt):  Use -ObjC flag with objc-edit.c

	* objcedit.c (round):  Made declaration global as well as definition.

	* objcedit.c (map_input):  Cast third arg to map_fd().

	* objcopt.c (get_objc):  Cast to (SEL *) two places.

	* objc-sel.m (_sel_writeHashTable):  Cast to (POHASH) two places.

	* objc-class.m (cache_flush, _cache_expand, _cache_flush):
	Made iteration variable unsigned.

	* hashtable.m (CopyIntoReadOnly):  Made iteration variable unsigned.

	* typedstream.m (InternalWriteObject):  Cleaned up printf format
	specifiers to match args.

	* typedstreamprivate.h:  Changed all function declarations to static.

	* objc-utils.m:  Cleaned up printf format specifiers to match args.

	* NXString.m (getBytes:range:, getCharacters:range:, NXFindCharacters,
	find:range:flags:usingTable:, _debugPrint):
	Made iteration variable unsigned.

	* NXCharacterSet.m (init, addCharacters:length:,
	removeCharacters:length:):  Made iteration variable unsigned.

Improvements to NXConstantString's (8/12/91):

	* NXString.[hm] (new):  Add version which raises exception.

	* NXString.[hm] (copyFromZone:):  Return an NXReadOnlyString copy
	instead of self, since our code might be unloaded.

	* NXString.[hm] (write:):  Write an NXReadOnlyString copy.

	* NXString.[hm] (finishUnarchiving):  Removed.

	* NXString.[hm] (free):  This now raises an error, since copyFromZone:
	no longer returns self.  You never allocate NXConstantString's, so you
	should never free them.


dpkg_scriptlib-1/python/dpkg/ar.py
# $Id: ar.py,v 1.1.1.1 1999/08/05 00:43:25 wsanchez Exp $
#
# Copyright (C) 1997  Klee Dienes <klee@mit.edu>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

import sys, string, subfile

def parse_string (s):
	e = len (s) - 1
	while ((e != -1) and (s[e] == ' ')):
		e = e - 1
	return s[0:e+1]

def parse_num (s):
	return string.atoi (s)

def parse_oct (s):
	return string.atoi (s, 8)

def parse_header (s):
	r = {}
	r['name'] = parse_string (s[0:16])
	r['date'] = parse_num (s[16:28])
	r['uid'] = parse_oct (s[28:34])
	r['gid'] = parse_oct (s[34:40])
	r['mode'] = parse_oct (s[40:48])
	r['size'] = parse_num (s[48:58])
	r['fmag'] = s[58:60]
	return r

def fseek (f, i):
	try:
		f.seek (i, 0)
		return
	except IOError, (errno, errstr):
		if (errno != 29):
			raise IOError, (errno, errstr)
	while (f.tell () < i):
		f.read (min (4096, i - f.tell()))
	if (i < f.tell()):
		raise ValueError, 'unable to seek backwards on stream'

def parse_file (f, p):
	s = f.read (8)
	if (s != '!<arch>\n'):
		raise ValueError, 'archive file is missing magic number (read "%s")' % s
	i = 8
	while 1:
		fseek (f, i)
		s = f.read (60)
		if (s == ''):
			return
		if (len (s) != 60):
			raise ValueError, 'bad read'
		r = parse_header (s)
		r['offset'] = i + 60
		i = i + 60 + (((r['size'] + 1) / 2) * 2)
		r['file'] = subfile.subfile (f, r['offset'], r['size'])
		p (f, r)
